compatibility_spark-3/src/main/scala/za/co/absa/hyperdrive/compatibility/impl/DeltaCDCToSCD2Writer.scala

```diff
@@ -0,0 +1,245 @@
+    dataFrame
+      .withColumn(
+        START_DATE_COLUMN,
+        col(configuration.timestampColumn)
+      )
+
+      .withColumn(
+        END_DATE_COLUMN,
+        functions.when(
+          col(MERGE_KEY_COLUMN).isNotNull.and(
+            lag(MERGE_KEY_COLUMN, 1, null).over(idWindowDesc).isNotNull
+          ),
+          col(END_DATE_COLUMN)
+        ).when(
+          col(MERGE_KEY_COLUMN).isNotNull.and(
+            lag(MERGE_KEY_COLUMN, 1, null).over(idWindowDesc).isNull
+          ).and(
+            col(configuration.timestampColumn).equalTo(
+              lag(s"${configuration.timestampColumn}", 1, null).over(idWindowDesc)
+            )
+          ),
+          lag(START_DATE_COLUMN, 2, null).over(idWindowDesc)
+        ).otherwise(
+          lag(START_DATE_COLUMN, 1, null).over(idWindowDesc)
+        )
+      )
+      .withColumn(
+        END_DATE_COLUMN,
+        functions
+          .when(col(configuration.operationColumn).equalTo(configuration.operationDeleteValue), col(START_DATE_COLUMN))
+          .when(col(configuration.operationColumn).notEqual(configuration.operationDeleteValue), col(END_DATE_COLUMN))
+          .otherwise(null)
+      )
+      .withColumn(
+        IS_CURRENT_COLUMN,
+        functions.when(col(END_DATE_COLUMN).isNull, lit(true)).otherwise(lit(false))
+      )
+  }
+
+  private def removeDuplicates(inputDF: DataFrame): DataFrame = {
+    val dataFrameWithSortColumns = getDataFrameWithSortColumns(inputDF, SORT_FIELD_PREFIX)
+
+    val originalFieldNames = inputDF.schema.fieldNames.mkString(",")
+    val sortColumnsWithPrefix = configuration.precombineColumns.map(precombineColumn =>
+      s"$SORT_FIELD_PREFIX$precombineColumn"
+    )
+
+    dataFrameWithSortColumns
+      .selectExpr(
+        s"${configuration.keyColumn}",
+        s"${configuration.timestampColumn}",
+        s"struct(${sortColumnsWithPrefix.mkString(",")}, $originalFieldNames) as otherCols"
+      )
+      .groupBy(s"${configuration.keyColumn}", s"${configuration.timestampColumn}")
+      .agg(functions.max("otherCols").as("latest"))
+      .filter(col("latest").isNotNull)
+      .withColumn("latest", new Column(AssertNotNull(col("latest").expr)))
+      .selectExpr("latest.*")
+      .drop(sortColumnsWithPrefix: _*)
+  }
+
+  private def getDataFrameWithSortColumns(dataFrame: DataFrame, sortFieldsPrefix: String): DataFrame = {
+    configuration.precombineColumns.foldLeft(dataFrame) { (df, precombineColumn) =>
+      val order = configuration.precombineColumnsCustomOrder.getOrElse(precombineColumn, Seq.empty[String])
+      order match {
+        case o if o.isEmpty =>
+          df.withColumn(s"$sortFieldsPrefix$precombineColumn", col(precombineColumn))
+        case o =>
+          val orderString = o.mkString(STRING_SEPARATOR)
+          df.withColumn(
+            s"$sortFieldsPrefix$precombineColumn",
+            functions.expr(s"""locate($precombineColumn, "$orderString")""")
```

Comment:
- Yes, that was my intent. If not defined it will have lowest precedence. Do you mean that it should have highest precedence ?
- I used array_position function instead. Fixed

Reply:
- Yeah, you're right. I mixed up the precedence